src/core/Hydrogen.cpp

```diff
@@ -945,17 +945,34 @@ inline void audioEngine_process_playNotes( unsigned long nframes )
 		// use this to support realtime events when not playing
 		framepos = pHydrogen->getRealtimeFrames();
 	}
-
+	
 	AutomationPath *vp = pSong->getVelocityAutomationPath();
 	
 
+	int nSongLength = 0;
+	if ( pSong->getMode() == Song::SONG_MODE ) {
+		nSongLength = pSong->lengthInTicks();
+	}
+
 	// reading from m_songNoteQueue
 	while ( !m_songNoteQueue.empty() ) {
 		Note *pNote = m_songNoteQueue.top();
 
 		float velocity_adjustment = 1.0f;
+		
+		// Velocity Adjustment Automation
 		if ( pSong->getMode() == Song::SONG_MODE ) {
-			float fPos = m_nSongPos + (pNote->get_position()%192) / 192.f;
+			// we need to calculate the total length (in ticks) of pattern columns BEFORE the playing column
+			int nPreviousColumnsLength;
+			 // findPatternInTick() writes the needed value into nPreviousColumnsLength
+			int tmp = findPatternInTick( m_nPatternStartTick, false, &nPreviousColumnsLength );
```

Comment:
This line does not seem to depend on the `Note`s  the while loop is iterating over. How about moving it in front of the loop?

Reply:
Are you sure  that notes can't be in different pattern columns?

Reply:
Ah ok, the function is called for every tick, (I thought it was for the audio buffer)

Reply:
I set some debug `printf()`  and I see that:
` audioEngine_process_playNotes() ` is called every nframes. Sometimes the notes in the loop belong to different pattern column.
So if I am not wrong now, I was wrong before when I used `m_nSongPos` trusting in its value. Instead that value has to be calculated as well in the loop

Reply:
> I set some debug printf() and I see that:
audioEngine_process_playNotes() is called every nframes. Sometimes the notes in the loop belong to different pattern columns.

Yes. The `audioEngine_process()` function is registered in the audio driver which in turns calls it on every `nframes` (buffer size) of audio that needs to be processed. After - and among other things - this function calls `audioEngine_updateNoteQueue`  (which sets `m_nSongPos` and `m_nPatternStartTick`) it calls `audioEngine_process_playNotes()`. These calls for handling the audio do live in a single thread and are called sequentially. Only by interaction with the GUI, by MIDI or OSC - which all live in different threads - other core routines can be run outside this sequential order. So, whether or not the Note belongs to a different column those two variables will be the same for all Notes in the note queue.

> Sometimes the notes in the loop belong to different pattern columns.

I haven't looked into this part too closely. But in general you can have Instruments with very long samples. A Note associated with them might play for several (short) patterns - e.g. a gong or a ride.

Reply:
> So, whether or not the Note belongs to a different column those two variables will be the same for all Notes in the note queue

So do you confirm that we cannot use `m_nSongPos` to calculate `velocity_adjustment` because `m_nSongPos` stores the starting position of the buffer and the buffer size may extend to the next pattern column in the song?

Reply:
> So do you confirm that we cannot use m_nSongPos to calculate velocity_adjustment because m_nSongPos stores the starting position of the buffer and the buffer size may extend to the next pattern column in the song?

I took a deeper look at the code. This should work:

```c++
	AutomationPath *vp = pSong->getVelocityAutomationPath();
	
	int nSongLength = 0;
	if ( pSong->getMode() == Song::SONG_MODE ) {
		nSongLength = pSong->lengthInTicks();
	}

	// reading from m_songNoteQueue
	while ( !m_songNoteQueue.empty() ) {
		Note *pNote = m_songNoteQueue.top();
		float velocity_adjustment = 1.0f;
		
		// Velocity Adjustment Automation
		if ( pSong->getMode() == Song::SONG_MODE ) {
			float fPos = static_cast<float>( m_nSongPos )
				+ ( static_cast<float>( pNote->get_position() % nSongLength - m_nPatternStartTick )
					/ static_cast<float>(pHydrogen->getCurrentPatternList()->longest_pattern_length()) );
			velocity_adjustment = vp->get_value(fPos);
		}

```

`m_nSongPos` is always the current pattern number and the second `findPatternInTick()` call can be omitted since `m_nPatternStartTick` and `nPreviousColumnsLength` will always be the same.

Reply:
Ok I tested and it works indeed. So some notes of the queue don't start in the current tick.

>`+ ( static_cast<float>( pNote->get_position() % nSongLength - m_nPatternStartTick )
`

I wanted to use a formula in the way that it could directly work for `double` positions and length ( without `%` )... but of course until it gets merged the result is the same.

Since the velocity automation is used only `if ( isNoteStart || isOldNote )` (the other notes in the queue are skipped if I understand now), can we put the velocity adjustment calculation into that scope?

Reply:
> I wanted to use a formula in the way that it could directly work for double positions and length ( without % )... but of course until it gets merged the result is the same.

Okay. But the equation should work for `double` positions too. The only thing that has to be adapted is to change the casts into `static_cast<double>` and the modulo operator `%` into the [std::fmod](https://en.cppreference.com/w/cpp/numeric/math/fmod) function

> Since the velocity automation is used only if ( isNoteStart || isOldNote ) (the other notes in the queue are skipped if I understand now), can we put the velocity adjustment calculation into that scope?

Sure